---
title: "Part 3 - B cell subtyping"
output: html_document
---

Single-cell analysis of human B cell maturation predicts how antibody class switching shapes selection dynamics In response to T cell-dependent antigens, mature B cells are stimulated to form germinal centers, which are the place of B cell affinity maturation and selection. Here, we dissected the heterogeneity of germinal center B cells and reconstructed their pathway of differentiation using single-cell transcriptomic analysis.

# Annotating the cell types using Human Protein Atlas

### load necessary libraries

```{r}
library(Seurat)
library(tidyverse)
library(dplyr)
library(RColorBrewer)
library(UCell)
```


read in the reference pbmc data obtained from 

```{r}
pbmc.ref <- readRDS("/Users/hawacoulibaly/Downloads/SEURAT_OBJECTS/HumanTonsil_BCells_scRNA_SeuratObject.rds")
pbmc.ref <- UpdateSeuratObject(pbmc.ref)
pbmc.ref
```

```{r}
# pbmc.44k <- subset(pbmc.ref, subset=Lineage %in% "B Cells")
```

```{r}
ncol(pbmc.ref)
```

```{r}
pbmc.ref@meta.data %>% head()
```
Next step discovers the most variable features (genes) in Bcell susets - these are usually most interesting for downstream analysis.

```{r}
pbmc.ref <- FindVariableFeatures(pbmc.ref, selection.method = "vst", nfeatures = 4000)
```

Identify the 10 most highly variable genes in Bcell subtypes:

```{r}
top10 <- head(VariableFeatures(pbmc.ref), 10)
top10
```

```{r}
plot1 <- VariableFeaturePlot(pbmc.ref)
LabelPoints(plot = plot1, points = top10, repel = TRUE, xnudge = 0, ynudge = 0)
```


```{r}
## how many cells for each cell type? 12 B  cell sub-types
table(pbmc.ref@meta.data$CellType)
```
Notice that only B cell subtypes are present

```{r}
# 10k cells
length(colnames(pbmc.ref))
```

```{r}
p1<- DimPlot(pbmc.ref, group.by = "CellType", label = TRUE, repel = TRUE) + NoLegend() + ggtitle("44k pbmc")
```


```{r}
# read in our 3k human PBMC data
pbmc<- readRDS("/Users/hawacoulibaly/Documents/pbmc_3k_annotated.RDS")

p2<- DimPlot(pbmc, group.by = "seurat_clusters", label = TRUE, repel = TRUE)+xlim(-15, 20)+ylim(-15, 5) + NoLegend() + ggtitle("3k pbmc")

CombinePlots(plots = list(p1, p2))
```
Now, we can identify anchors between the two dataset and use these anchors to transfer the celltype labels we learned from the 10K scRNA-seq data to the 3k pmbc cells.

```{r}
transfer.anchors <- FindTransferAnchors(reference = pbmc.ref, query = pbmc, features = VariableFeatures(object = pbmc.ref), 
    reference.assay = "RNA", query.assay = "RNA", reduction = "pcaproject")
```

Note if transferring scRNAseq label to scATACseq data, set reduction = “cca” is recommended

To transfer the cluster ids, we provide a vector of previously annotated cell type labels for the RNA to the refdata parameter. The output will contain a matrix with predictions and confidence scores for each 44k pbmc cell.


```{r}
ElbowPlot(pbmc.ref)
```


```{r}
celltype.predictions <- TransferData(anchorset = transfer.anchors, refdata = pbmc.ref$CellType, dims = 1:10)
```
```{r}
head(celltype.predictions)
```

```{r}
pbmc<- AddMetaData(pbmc, metadata = celltype.predictions)

pbmc@meta.data %>% head()
```

```{r}
DimPlot(pbmc, group.by = "predicted.id", label = TRUE, repel = TRUE)+xlim(-15, 20)+ylim(-15, 7) + NoLegend() + ggtitle("3k pbmc")
```
confidence of the label transfering


```{r}
pbmc@meta.data$prediction.score.max %>%
  tibble::enframe() %>%
  ggplot(aes(x=value)) + 
  geom_histogram(color = "white", bins = 50) +
  geom_vline(xintercept = 0.9, linetype = 2, color = "red") + 
  theme_classic(base_size = 14) +
  ggtitle("distribution of max prediciton score")
```

